File: Python program/Source code/Source_code.py
# Program to sort an integer array using 4 algorithms: Counting Sort, Heap Sort, Odd-Even Sort, Pigeonhole Sort.
# Instructions:
# 1. Enter the number of elements (n) for the array (must be a non-negative integer).
# 2. Enter n integers one by one to create the array.
# 3. Choose a sorting algorithm from the menu by entering a number from 1 to 5:
#    - 1: Sort using Counting Sort
#    - 2: Sort using Heap Sort
#    - 3: Sort using Odd-Even Sort
#    - 4: Sort using Pigeonhole Sort
#    - 5: Exit the program
# 4. After selection, the program will display the sorted array.
# 5. You can choose another algorithm or exit.
# Note: Entering incorrect formats (e.g., letters instead of numbers) will cause an error and display a message.

def counting_sort(arr):
    n = len(arr)
    if n == 0:
        return []

    MAX = max(arr)
    MIN = min(arr)

    range_of_elements = MAX - MIN + 1
    count = [0] * range_of_elements

    for i in range(n):
        count[arr[i] - MIN] += 1

    for i in range(1, range_of_elements):
        count[i] += count[i - 1]

    sorted_arr = [0] * n
    for i in range(n - 1, -1, -1):
        sorted_arr[count[arr[i] - MIN] - 1] = arr[i]
        count[arr[i] - MIN] -= 1

    return sorted_arr

def heapify(arr, n, i):
    largest = i
    left = 2 * i + 1
    right = 2 * i + 2

    if left < n and arr[left] > arr[largest]:
        largest = left
    if right < n and arr[right] > arr[largest]:
        largest = right

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)

def heap_sort(arr):
    n = len(arr)
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)
    for i in range(n - 1, 0, -1):
        arr[0], arr[i] = arr[i], arr[0]
        heapify(arr, i, 0)
    return arr

def odd_even_sort(arr):
    n = len(arr)
    is_sorted = False
    while not is_sorted:
        is_sorted = True
        for i in range(1, n - 1, 2):
            if arr[i] > arr[i + 1]:
                arr[i], arr[i + 1] = arr[i + 1], arr[i]
                is_sorted = False
        for i in range(0, n - 1, 2):
            if arr[i] > arr[i + 1]:
                arr[i], arr[i + 1] = arr[i + 1], arr[i]
                is_sorted = False
    return arr

def pigeonhole_sort(arr):
    if not arr:
        return arr
    MIN = min(arr)
    MAX = max(arr)
    range_size = MAX - MIN + 1

    holes = [0] * range_size

    for num in arr:
        holes[num - MIN] += 1

    index = 0
    for i in range(range_size):
        while holes[i] > 0:
            arr[index] = i + MIN
            index += 1
            holes[i] -= 1
    return arr

def display_menu():
    print("\n=== SORTING MENU ===")
    print("1. Counting Sort")
    print("2. Heap Sort")
    print("3. Odd-Even Sort")
    print("4. Pigeonhole Sort")
    print("5. Exit")

def main():
    # Input array
    try:
        n = int(input("Enter the number of elements: "))
        if n < 0:
            print("Number of elements cannot be negative!")
            return
        arr = []
        for i in range(n):
            arr.append(int(input(f"Enter element {i+1}: ")))
        print("Original array:", arr)

        while True:
            display_menu()
            choice = input("Choose an algorithm (1-5): ")

            # Create a copy of the array to preserve the original
            arr_copy = arr.copy()
            sorted_arr = []

            if choice == '1':
                sorted_arr = counting_sort(arr_copy)
                print("Array sorted by Counting Sort:", sorted_arr)
            elif choice == '2':
                sorted_arr = heap_sort(arr_copy)
                print("Array sorted by Heap Sort:", sorted_arr)
            elif choice == '3':
                sorted_arr = odd_even_sort(arr_copy)
                print("Array sorted by Odd-Even Sort:", sorted_arr)
            elif choice == '4':
                sorted_arr = pigeonhole_sort(arr_copy)
                print("Array sorted by Pigeonhole Sort:", sorted_arr)
            elif choice == '5':
                print("Program terminated!")
                break
            else:
                print("Invalid choice! Please choose again.")
    except ValueError:
        print("Error: Please enter integers only!")

if __name__ == "__main__":
    main()

File: C++ Program/Source-Code.cpp
#include <iostream>
#include <vector>
#include <algorithm>

std::vector<int> countingSort(const std::vector<int>& arr) {
    int n = arr.size();
    int MAX = arr[0], MIN = arr[0];

    for (int i = 0; i < n; ++i) {
        MAX = std::max(MAX, arr[i]);
        MIN = std::min(MIN, arr[i]);
    }

    int range = MAX - MIN + 1;
    std::vector<int> count(range, 0);

    for (int i = 0; i < n; ++i)
        ++count[arr[i] - MIN];
    for (int i = 1; i < range; ++i)
        count[i] += count[i - 1];

    std::vector<int> sorted(n);

    for (int i = n - 1; i >= 0; --i) {
        --count[arr[i] - MIN];
        sorted[count[arr[i] - MIN]] = arr[i];
    }

    return sorted;
}

void heapify(std::vector<int>& arr, int i, int n) {
    int largest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;

    if (left < n && arr[left] > arr[largest])
        largest = left;
    if (right < n && arr[right] > arr[largest])
        largest = right;

    if (largest != i) {
        std::swap(arr[i], arr[largest]);
        heapify(arr, largest, n);
    }
}

void heapSort(std::vector<int>& arr) {
    int n = arr.size();
    for (int i = n / 2 - 1; i >= 0; --i)
        heapify(arr, i, n);
    for (int i = n - 1; i > 0; --i) {
        std::swap(arr[0], arr[i]);
        heapify(arr, 0, i);
    }
}

void oddEvenSort(int arr[], int n) {
    bool isSorted = false;
    while (!isSorted) {
        isSorted = true;
        for (int i = 1; i < n - 1; i += 2) {
            if (arr[i] > arr[i + 1]) {
                std::swap(arr[i], arr[i + 1]);
                isSorted = false;
            }
        }
        for (int i = 0; i < n - 1; i += 2) {
            if (arr[i] > arr[i + 1]) {
                std::swap(arr[i], arr[i + 1]);
                isSorted = false;
            }
        }
    }
}

void pigeonholeSort(int arr[], int n) {
    int MIN = arr[0], MAX = arr[0];

    for (int i = 1; i < n; i++) {
        MIN = std::min(MIN, arr[i]);
        MAX = std::max(MAX, arr[i]);
    }

    int range = MAX - MIN + 1;
    std::vector<int> holes(range, 0);

    for (int i = 0; i < n; i++)
        ++holes[arr[i] - MIN];

    int index = 0;
    for (int i = 0; i < range; i++) {
        while (holes[i]--) {
            arr[index] = i + MIN;
            ++index;
        }
    }
}


int main() {
    int n;
    std::cout << "Enter the number of elements: ";
    std::cin >> n;

    std::vector<int> vec(n);
    int* arr = new int[n];
    std::cout << "Enter " << n << " elements: ";
    for (int i = 0; i < n; ++i) {
        std::cin >> vec[i];
        arr[i] = vec[i];
    }

    while (true) {
        std::cout << "\n=== SORTING MENU ===\n";
        std::cout << "1. Sort using Counting Sort\n";
        std::cout << "2. Sort using Heap Sort\n";
        std::cout << "3. Sort using Odd-Even Sort\n";
        std::cout << "4. Sort using Pigeonhole Sort\n";
        std::cout << "5. Exit the program\n";
        std::cout << "Choose (1-5): ";

        int choice;
        std::cin >> choice;

        if (choice == 5) break;

        std::vector<int> vecCopy = vec;
        int* arrCopy = new int[n];
        for (int i = 0; i < n; ++i) arrCopy[i] = arr[i];

        switch (choice) {
        case 1: {
            std::vector<int> sorted = countingSort(vecCopy);
            std::cout << "Result (Counting Sort): ";
            for (int x : sorted) std::cout << x << " ";
            break;
        }
        case 2: {
            heapSort(vecCopy);
            std::cout << "Result (Heap Sort): ";
            for (int x : vecCopy) std::cout << x << " ";
            break;
        }
        case 3: {
            oddEvenSort(arrCopy, n);
            std::cout << "Result (Odd-Even Sort): ";
            for (int i = 0; i < n; ++i) std::cout << arrCopy[i] << " ";
            break;
        }
        case 4: {
            pigeonholeSort(arrCopy, n);
            std::cout << "Result (Pigeonhole Sort): ";
            for (int i = 0; i < n; ++i) std::cout << arrCopy[i] << " ";
            break;
        }
        default:
            std::cout << "Invalid choice!\n";
        }
        std::cout << std::endl;
        delete[] arrCopy;
    }

    delete[] arr;
    return 0;
}

/*
Program to sort an array of integers using 4 algorithms: Counting Sort, Heap Sort, Odd-Even Sort, Pigeonhole Sort.
Usage Guide:

1. Enter the number of elements (n) in the array (must be a non-negative integer).
2. Enter n integers to create the array.
3. Choose a sorting algorithm from the menu by entering a number from 1 to 5:
   - 1: Sort using Counting Sort
   - 2: Sort using Heap Sort
   - 3: Sort using Odd-Even Sort
   - 4: Sort using Pigeonhole Sort
   - 5: Exit the program
After selecting, the program will display the sorted array.
4. You can choose another algorithm or exit.
Note: Entering incorrect formats (e.g., letters instead of numbers) will cause errors and display a message.
*/